agent/provider/src/execution/task_runner.rs

```diff
@@ -138,14 +138,15 @@ pub struct TaskRunner {
 }
 
 impl TaskRunner {
-    pub fn new(
+    pub fn new<P: AsRef<Path>>(
         client: ActivityProviderApi,
         config: TaskRunnerConfig,
         registry: ExeUnitsRegistry,
+        data_dir: P,
```

Comment:
What is the benefit of using `AsRef<Path>`, rather than just `&Path`?

Reply:
it is more generic
You can input both `Path` and `&Path` and `PathBuf` and `&PathBuf` as a input. Namely all that implements `AsRef<Path>`

Reply:
I understand that there is a trade-off between
```rust
fn foo(path: &Path) { // <- simpler function
    use(path);
}

let path_buf: PathBuf = ...;
foo(path_buf.as_ref());
```
and
```rust
fn foo<P: AsRef<Path>>(path: P) {
    let path = path.as_ref();
    use(path);
}

let path_buf: PathBuf = ...;
foo(path_buf); // <- simpler calling
```

I just think that the complexity function in second example is bigger than complexity of calling it in first example. Also using such templates might make compilation longer and output binary bigger (based on experience from C++ templates).

Reply:
pls read my answer carefully ;)
in template version `foo<P: AsRef<Path>>(path: P)` you can also do
```
let p : Path =...
foo(p)
```

and **any** other type that implements AsRef for Path

Reply:
You can't make `let p: Path = ...`. https://doc.rust-lang.org/std/path/struct.Path.html says
> This is an unsized type, meaning that it must always be used behind a pointer like & or Box. For an owned version of this type, see PathBuf.

But, let's take a `PathBuf` as an example:
```rust
fn foo(path: &Path) {}

let p: PathBuf = ...;
foo(&p);
```
This works just fine.

So can you provide a complete convincing example, where your version has a clear advantage?

Reply:
@etam Please take a look at https://doc.rust-lang.org/std/path/struct.Path.html#trait-implementations